File: api-gateway/internal/handler/parse.go
package handler

import (
	"context"
	"log"
	"time"

	"github.com/gin-gonic/gin"

	"vasset/api-gateway/internal/models"
	pb "vasset/api-gateway/proto"
)

// ParseHandler 解析处理器
type ParseHandler struct {
	parserClient pb.ParserServiceClient
	timeout      time.Duration
}

// NewParseHandler 创建解析处理器
func NewParseHandler(parserClient pb.ParserServiceClient, timeout time.Duration) *ParseHandler {
	return &ParseHandler{
		parserClient: parserClient,
		timeout:      timeout,
	}
}

// ParseURL 解析视频 URL
func (h *ParseHandler) ParseURL(c *gin.Context) {
	var req models.ParseRequest
	if err := c.ShouldBindJSON(&req); err != nil {
		models.BadRequest(c, "invalid request: "+err.Error())
		return
	}

	ctx, cancel := context.WithTimeout(c.Request.Context(), h.timeout)
	defer cancel()

	resp, err := h.parserClient.ParseURL(ctx, &pb.ParseURLRequest{
		Url:       req.URL,
		SkipCache: req.SkipCache,
	})
	if err != nil {
		models.InternalError(c, "failed to parse URL: "+err.Error())
		return
	}

	// 调试日志：从 gRPC 接收的格式
	log.Printf("[DEBUG-ParseHandler] Received from gRPC: %d formats", len(resp.Formats))

	// 转换格式列表
	formats := make([]models.VideoFormat, 0, len(resp.Formats))
	var maxHeight int32
	var videoCount, audioCount int
	for i, f := range resp.Formats {
		// 统计
		if f.Height > maxHeight {
			maxHeight = f.Height
		}
		if f.VideoCodec != "" && f.VideoCodec != "none" {
			videoCount++
		}
		if f.AudioCodec != "" && f.AudioCodec != "none" && (f.VideoCodec == "" || f.VideoCodec == "none") {
			audioCount++
		}

		// 调试前3个格式
		if i < 3 {
			log.Printf("[DEBUG-ParseHandler] Format[%d]: id=%s, height=%d, video_codec=%s, audio_codec=%s, filesize=%d",
				i, f.FormatId, f.Height, f.VideoCodec, f.AudioCodec, f.Filesize)
		}

		formats = append(formats, models.VideoFormat{
			FormatID:   f.FormatId,
			Quality:    f.Quality,
			Extension:  f.Extension,
			Filesize:   f.Filesize,
			Height:     f.Height,
			Width:      f.Width,
			FPS:        f.Fps,
			VideoCodec: f.VideoCodec,
			AudioCodec: f.AudioCodec,
			VBR:        f.Vbr,
			ABR:        f.Abr,
			ASR:        f.Asr,
		})
	}

	log.Printf("[DEBUG-ParseHandler] Sending to frontend: %d formats (video=%d, audio=%d, maxHeight=%d)",
		len(formats), videoCount, audioCount, maxHeight)

	models.Success(c, models.ParseResponse{
		VideoID:     resp.VideoId,
		Platform:    resp.Platform,
		Title:       resp.Title,
		Description: resp.Description,
		Duration:    resp.Duration,
		Thumbnail:   resp.Thumbnail,
		Author:      resp.Author,
		UploadDate:  resp.UploadDate,
		ViewCount:   resp.ViewCount,
		Formats:     formats,
	})
}

File: frontend-service/lib/api/parse.ts
import apiClient from '../api-client';
import { formatDuration } from '../format';

export interface VideoFormat {
    format_id: string;
    quality: string;
    extension: string;
    filesize: number;
    height: number;
    width?: number;
    fps: number;
    video_codec: string;
    audio_codec: string;
    vbr?: number;        // 视频码率 (kbps)
    abr?: number;        // 音频码率 (kbps)
    tbr?: number;        // 总码率 (kbps)
    asr?: number;        // 音频采样率 (Hz)
    format_note?: string; // 格式说明，如 "1080p", "medium"
}

export interface VideoInfo {
    video_id: string;
    platform: string;
    title: string;
    description: string;
    duration: number;        // 后端返回秒数
    durationFormatted: string; // 前端格式化后
    thumbnail: string;
    author: string;
    upload_date: string;
    view_count: number;
    formats: VideoFormat[];
    url: string;             // 前端自行保存
}

export const parseApi = {
    // 解析URL
    parseUrl: async (url: string, skipCache = false): Promise<VideoInfo> => {
        const response = await apiClient.post('/api/v1/parse', { url, skip_cache: skipCache });
        const data = response.data;

        // 调试日志
        console.log('[DEBUG-parseUrl] Received from API:', {
            formatsCount: data.formats?.length || 0,
            videoFormats: data.formats?.filter((f: any) => f.video_codec && f.video_codec !== 'none').length || 0,
            audioFormats: data.formats?.filter((f: any) => f.audio_codec && f.audio_codec !== 'none' && (!f.video_codec || f.video_codec === 'none')).length || 0,
            maxHeight: Math.max(...(data.formats?.map((f: any) => f.height || 0) || [0])),
            first3Formats: data.formats?.slice(0, 3).map((f: any) => ({
                format_id: f.format_id,
                height: f.height,
                video_codec: f.video_codec,
                audio_codec: f.audio_codec,
                filesize: f.filesize
            }))
        });

        return {
            ...data,
            durationFormatted: formatDuration(data.duration),
            url, // 保存原始URL
        };
    },
};
